Fitbit.Portable/FitBitHttpClientMessageHandler.cs

```diff
@@ -1,38 +1,59 @@
 ﻿namespace Fitbit.Api.Portable
 {
+    using System;
     using System.Diagnostics;
     using System.Net.Http;
     using System.Threading;
     using System.Threading.Tasks;
 
-    public class FitbitHttpClientMessageHandler : DelegatingHandler
+    internal class FitbitHttpClientMessageHandler : DelegatingHandler
     {
-        private IFitbitClientInterceptor logger;
+        private IFitbitInterceptor interceptor;
+        Func<Task<HttpResponseMessage>, CancellationToken, Task<HttpResponseMessage>> responseHandler;
 
-        public FitbitHttpClientMessageHandler(IFitbitClientInterceptor logger)
+        //private ITokenManager tokenManager;
+
+        public FitbitHttpClientMessageHandler(IFitbitInterceptor interceptor)
         {
-            this.logger = logger;
+            this.interceptor = interceptor;
+            //this.tokenManager = tokenManager;
+            responseHandler = ResponseHandler; 
             //Define the inner must handler. Otherwise exception is thrown.
             InnerHandler = new HttpClientHandler();
         }
 
+        //Handle the following method with EXTREME care as it will be invoked on ALL requests made by FitbitClient
         // We override the SendAsync method to intercept both the request and response path
         protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            Task<HttpResponseMessage> interceptorResponse = null;
             Debug.WriteLine("Entering Http client's request message handler. Request details: {0}", request.ToString());
-            if (logger != null)
-                logger.InterceptRequest(request, cancellationToken);
-
-            return base.SendAsync(request, cancellationToken).ContinueWith(
-                     requestTask =>
-                     {
-                         Debug.WriteLine("Entering Http client's response message handler. Response details: {0}", requestTask.Result);
-                         if (logger != null)
-                             logger.InterceptResponse(requestTask.Result, cancellationToken);
-
-                         return requestTask.Result;
-                     }, TaskContinuationOptions.OnlyOnRanToCompletion
+
+            if (interceptor != null)
+                interceptorResponse = interceptor.InterceptRequest(request, cancellationToken);
+
+            if(interceptorResponse != null) //then highjack the request pipeline and return the HttpResponse returned by interceptor. Invoke Response handler at return.
```

Comment:
I don't like the "highjack" of the whole request pipeline, it removes the possibility of people plugging in more things over time and knowing what it's actually doing. I've implemented a message handler pipeline so it's plug and play for future enhancements, let me know what you think - https://github.com/WestDiscGolf/Fitbit.NET/commit/c92fa0002914f12b23cf28566e1c824bee597561